metalcloud/resource_metalcloud_extension_instance.go

```diff
@@ -0,0 +1,586 @@
+
+		if r.StatusCode == http.StatusNotFound {
+			return nil, errNotFound
+		}
+
+		return nil, err
+	}
+
+	return &i, err
+}
+
+func getExtension(op *opCtx, id int) (*sdk2.ExtensionDto, error) {
+	x, r, err :=
+		op.client.ExtensionApi.GetExtension(op.ctx, float64(id))
+	if err != nil {
+		op.logf("%v: %v\r\n", r.StatusCode, err)
+
+		if r.StatusCode == http.StatusNotFound {
+			return nil, errNotFound
+		}
+
+		return nil, err
+	}
+
+	return &x, err
+}
+
+type findExtensionInstanceByLabelAndInfrastructure struct {
+	InfrastructureId string
+	Label            string
+}
+
+func (f *findExtensionInstanceByLabelAndInfrastructure) ToString() string {
+	//prefix := "filter."
+	prefix := ""
+	return fmt.Sprintf("%v%v=%v,%v%v=%v", prefix, "infrastructureId", f.InfrastructureId, prefix, "label", f.Label)
+}
+
+func findExtensionInstance(op *opCtx, infrastructureId int, label string) (*sdk2.ExtensionInstanceDto, error) {
+	filter := &findExtensionInstanceByLabelAndInfrastructure{
+		InfrastructureId: fmt.Sprintf("$eq:%v", infrastructureId),
+		Label:            fmt.Sprintf("$eq:%v", label),
+	}
+
+	find := new(sdk2.ExtensionInstanceApiGetExtensionInstancesOpts)
+	find.Filter = optional.NewInterface(filter.ToString())
+	find.Page = optional.NewInt32(0)
+	find.Limit = optional.NewInt32(2)
+
+	x, _, err :=
+		op.client.ExtensionInstanceApi.GetExtensionInstances(op.ctx, find)
+	if err != nil {
+		op.logf("%v\r\n", err)
+
+		return nil, err
+	}
+
+	if len(x.ExtensionInstances) == 1 {
+		return &(x.ExtensionInstances[0]), nil
+	}
+
+	return nil, nil
+}
+
+var errNotFound = errors.New("not found")
+
+func getExtensionInstance(op *opCtx, id int) (*sdk2.ExtensionInstanceDto, error) {
+	instance, r, err :=
+		op.client.ExtensionInstanceApi.GetExtensionInstance(op.ctx, float64(id))
+	if err != nil {
+		op.logf("%v: %v\r\n", r.StatusCode, err)
+
+		if r.StatusCode == http.StatusNotFound {
+			return nil, errNotFound
+		}
+
+		return nil, err
+	}
+
+	return &instance, nil
+}
+
+func createExtensionInstance(op *opCtx, dto *sdk2.CreateExtensionInstanceDto, infrastructure_id int) (*sdk2.ExtensionInstanceDto, error) {
+	instance, r, err :=
+		op.client.ExtensionInstanceApi.CreateExtensionInstance(op.ctx, *dto, float64(infrastructure_id))
+	if err != nil {
+		op.logf("%v: %v\r\n", r.StatusCode, err)
+
+		return nil, err
+	}
+
+	return &instance, nil
+}
+
+func updateExtensionInstance(op *opCtx, dto *sdk2.UpdateExtensionInstanceDto, id int) (*sdk2.ExtensionInstanceDto, error) {
+	instance, r, err :=
+		op.client.ExtensionInstanceApi.UpdateExtensionInstance(op.ctx, *dto, float64(id))
+	if err != nil {
+		op.logf("%v: %v\r\n", r.StatusCode, err)
+
+		return nil, err
+	}
+	return &instance, nil
+}
+
+func deleteExtensionInstance(op *opCtx, id int) error {
+	r, err :=
+		op.client.ExtensionInstanceApi.DeleteExtensionInstance(op.ctx, float64(id))
+	if err != nil {
+		op.logf("%v: %v\r\n", r.StatusCode, err)
+	}
+	return err
+}
+
+func op(ctx context.Context) (*opCtx, error) {
```

Comment:
Instead of creating new type, use context.WithValue to create context that includes additional values.
The diag.Diagnostics in my opinion should not be shared and should be instead returned by each function and tested by the calling function.